File: essdee/essdee/report/party_ledger/party_ledger.py
# Copyright (c) 2013, Aerele Technologies Pvt Ltd and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
import requests, json
from frappe.utils import today, cint, get_site_path
from frappe.www.printview import get_print_style
import base64
from bs4 import BeautifulSoup

def execute(filters=None):
	columns, data = [], []
	if not 'customer' in filters:
		return columns, data
	else:
		filters['party'] = [filters['customer']]
		filters.pop('customer')
	return get_gl_entries_via_api(filters)

def get_gl_entries_via_api(filters):
	host = '127.0.0.1'
	site_name = frappe.local.site
	port = frappe.get_site_config().webserver_port
	api_key = frappe.db.get_value("User", "Administrator", "api_key")
	api_secret = frappe.utils.password.get_decrypted_password("User", "Administrator", fieldname="api_secret")
	headers = {
		'Accept': 'application/json',
		'Authorization': f'token {api_key}:{api_secret}',
		'Content-Type': 'application/json',
		'X-Forwarded-For':'127.0.0.1',
		'X-Forwarded-Proto': 'https://',
		'X-Frappe-Site-Name': site_name,
		'Host': site_name
	}
	url = f'http://{host}:{port}/api/method/essdee.essdee.report.party_ledger.party_ledger.get_gl_entries'
	payload = {
		'filters': filters
	}
	response = json.loads(requests.request("POST", url, headers=headers, data = json.dumps(payload)).text.encode('utf8'))['message']
	if len(response) == 2:
		columns, data  = response[0], response[1]
	for column in columns:
		if 'fieldtype' in column:
			if column['fieldtype'] == 'Link' or column['fieldtype'] == 'Dynamic Link':
				column['fieldtype'] = 'Data'
				del column['options']
	return columns, data

@frappe.whitelist()
def get_gl_entries(filters):
	columns, data = [], []
	from erpnext.accounts.report.general_ledger.general_ledger import execute as execute_general_ledger
	columns, data = execute_general_ledger(frappe._dict(filters))
	return columns, data

@frappe.whitelist()
def get_custom_ledger_list(customer, list_limit):
	list_limit = cint(list_limit)
	default_company = frappe.defaults.get_user_default("Company")
	abbr = frappe.db.get_value("Company", default_company, "abbr")
	ledger_filters = {
		'company': default_company,
		'from_date': today(),
		'to_date': today(),
		'party_type': 'Customer',
		'group_by': 'Group by Voucher (Consolidated)',
		'presentation_currency': 'INR',
		'party': [customer]}
	col, data = get_gl_entries(ledger_filters)
	running_bal = data[-1].balance
	ledger_list = []
	common_query = """select posting_date,voucher_no,voucher_type,sum(credit),sum(debit),against
		from `tabGL Entry`where docstatus=1 and company=%(company)s and 
		party_type=%(party_type)s and party=%(party)s 
		group by voucher_no 
		order by posting_date desc,voucher_no desc """
	common_query_values = {"company": default_company, "party_type": "Customer", "party": customer}
	if list_limit:
		common_query += 'limit %(list_limit)s'
		common_query_values.update(list_limit=list_limit)
	gl_entry = frappe.db.sql(common_query, common_query_values, as_list=1)

	if gl_entry:
		for entry in gl_entry:
			ledger_list.append({
				"date": entry[0].strftime('%d-%m-%Y'),
				"voucher_no": entry[1],
				"voucher_type": entry[2],
				"amount": round(abs(entry[3]-entry[4]),2),
				"against_account": entry[5].replace(f' - {abbr}', ''),
				"balance": running_bal
			})
			running_bal = (running_bal + entry[3]) if entry[3] else (running_bal - entry[4])
	return {'list': ledger_list}

@frappe.whitelist()
def get_custom_invoice_list(customer, list_limit):
	list_limit = cint(list_limit)
	custom_invoice_list = []
	common_dict = dict(fields={'posting_date',
				'name',
				'grand_total',
				'packing_slip_number',
				'lr_number',
				'lr_date',
				'transport',
				'order_number',
				'number_of_cartons',
				'destination'},
				filters={"customer":customer, "docstatus": 1}, order_by='posting_date desc, name desc')
	if list_limit:
		common_dict.update(limit = list_limit)
	invoice_list = frappe.get_list('Sales Invoice', **common_dict)
	if invoice_list:
		for entry in invoice_list:
			custom_invoice_list.append({
				"date": entry['posting_date'].strftime('%d-%m-%Y'),
				"voucher_no": entry['name'],
				"amount": entry['grand_total'],
				"cashback_claim_status": "NA",
				"packing_slip_no": entry['packing_slip_number'],
				"lr_no": entry['lr_number'],
				"lr_date": entry['lr_date'].strftime('%d-%m-%Y') if entry['lr_date'] else None,
				"transport": entry['transport'],
				"order_no": entry['order_number'],
				"no_of_cartons": entry['number_of_cartons'],
				"destination": entry['destination']
			})
	return {'list': custom_invoice_list}

@frappe.whitelist()
def get_accounts_document_print_html(customer, doctype, doc_name, print_format=None, no_letterhead=0):
	if doctype == "Sales Invoice" and frappe.db.get_value(doctype, doc_name, "customer") != customer:
		current_user = frappe.get_user().name
		frappe.log_error("docname: "+doc_name+"doctype: "+doctype+" | user: "+current_user, "View Invoice PDF Error")
		frappe.throw('Unauthorized customer to view corresponding pdf')

	if print_format is None:
		print_format = frappe.get_meta(doctype).default_print_format
	html=frappe.get_print(doctype, doc_name, print_format, no_letterhead=no_letterhead)
	return image_conversion(html)

@frappe.whitelist()
def get_ledger_print_html(customer=None, from_date=None, to_date=None, ledger_filters=None):
	if not ledger_filters:
		default_company = frappe.defaults.get_user_default("Company")
		ledger_filters = {
			'company': default_company,
			'from_date': from_date,
			'to_date': to_date,
			'party_type': 'Customer',
			'group_by': 'Group by Voucher (Consolidated)',
			'presentation_currency': 'INR',
			'party': [customer]}
	if isinstance(ledger_filters, str):
		ledger_filters = eval(ledger_filters)
	if 'customer' in ledger_filters:
		ledger_filters['party'] = [ledger_filters['customer']]
		ledger_filters.pop('customer')
	if 'party' in ledger_filters:
		col, data = get_gl_entries_via_api(ledger_filters)
		letterhead_doc = frappe.get_doc("Letter Head", {"is_default": 1})
		ledger_filters['letter_head'] = letterhead_doc.content
		base_template_path = "frappe/www/printview.html"
		template_path = "essdee/essdee/report/party_ledger/party_ledger.html"
		if data:
			html = frappe.render_template(template_path, \
				{"filters": ledger_filters, "data": data})
			html = frappe.render_template(base_template_path, {"body": html, \
				"css": get_print_style(), "title": "Statement For " + ledger_filters['party'][0]})
			return image_conversion(html)

def image_conversion(html):
	soup = BeautifulSoup(html)
	for img in soup.find_all('img'):
		img_url = img.get('src')
		img_path = img_url.split('/')
		if img_path[1] == "private":
			img_url = get_site_path(*img_path)
		elif img_path[1] == "files":
			img_url = get_site_path("public", *img_path)
		else:
			continue
		ext = img_url.split('.')[-1]
		encoded = base64.b64encode(open(img_url, 'rb').read()).decode('utf-8')
		prefix = f'data:image/{ext};base64,'
		img.attrs['src'] = prefix + encoded
	return (str(soup))

File: essdee/essdee/report/party_ledger/party_ledger.js
// Copyright (c) 2016, Aerele Technologies Pvt Ltd and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Party Ledger"] = {
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldtype": "Break",
		},
		{
			"fieldname":"finance_book",
			"label": __("Finance Book"),
			"fieldtype": "Link",
			"options": "Finance Book",
			"hidden": 1
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1,
			"width": "60px"
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
			"width": "60px"
		},
		{
			"fieldtype": "Break",
		},
		{
			"fieldname":"account",
			"label": __("Account"),
			"fieldtype": "Link",
			"options": "Account",
			"get_query": function() {
				var company = frappe.query_report.get_filter_value('company');
				return {
					"doctype": "Account",
					"filters": {
						"company": company,
					}
				}
			},
			"hidden": 1
		},
		{
			"fieldname":"voucher_no",
			"label": __("Voucher No"),
			"fieldtype": "Data",
			on_change: function() {
				frappe.query_report.set_filter_value('group_by', "Group by Voucher (Consolidated)");
			},
			"hidden": 1	
		},
		{
			"fieldname":"party_type",
			"label": __("Party Type"),
			"fieldtype": "Link",
			"options": "Party Type",
			"default": "Customer",
			on_change: function() {
				frappe.query_report.set_filter_value('party', "");
			},
			"hidden": 1
		},
		{
			"fieldname":"customer",
			"label": __("Customer"),
			"fieldtype": "Link",
			"options": "Customer",
			on_change: () => {
				var customer = frappe.query_report.get_filter_value('customer');
				var company = frappe.query_report.get_filter_value('company');
				if (customer) {
					frappe.db.get_value('Customer', customer, ["tax_id", "customer_name", "payment_terms"], function(value) {
						frappe.query_report.set_filter_value('tax_id', value["tax_id"]);
						frappe.query_report.set_filter_value('customer_name', value["customer_name"]);
						frappe.query_report.set_filter_value('payment_terms', value["payment_terms"]);
					});

					frappe.db.get_value('Customer Credit Limit', {'parent': customer, 'company': company},
						["credit_limit"], function(value) {
						if (value) {
							frappe.query_report.set_filter_value('credit_limit', value["credit_limit"]);
						}
					}, "Customer");
				} else {
					frappe.query_report.set_filter_value('tax_id', "");
					frappe.query_report.set_filter_value('customer_name', "");
					frappe.query_report.set_filter_value('credit_limit', "");
					frappe.query_report.set_filter_value('payment_terms', "");
				}
			}
		},
		{
			"fieldname":"party",
			"label": __("Party"),
			"fieldtype": "MultiSelectList",
			"hidden": 1,
			get_data: function(txt) {
				if (!frappe.query_report.filters) return;

				let party_type = frappe.query_report.get_filter_value('party_type');
				if (!party_type) return;

				return frappe.db.get_link_options(party_type, txt);
			},
			on_change: function() {
				var party_type = frappe.query_report.get_filter_value('party_type');
				var parties = frappe.query_report.get_filter_value('party');

				if(!party_type || parties.length === 0 || parties.length > 1) {
					frappe.query_report.set_filter_value('party_name', "");
					frappe.query_report.set_filter_value('tax_id', "");
					return;
				} else {
					var party = parties[0];
					var fieldname = erpnext.utils.get_party_name(party_type) || "name";
					frappe.db.get_value(party_type, party, fieldname, function(value) {
						frappe.query_report.set_filter_value('party_name', value[fieldname]);
					});

					if (party_type === "Customer" || party_type === "Supplier") {
						frappe.db.get_value(party_type, party, "tax_id", function(value) {
							frappe.query_report.set_filter_value('tax_id', value["tax_id"]);
						});
					}
				}
			}
		},
		{
			"fieldname":"party_name",
			"label": __("Party Name"),
			"fieldtype": "Data",
			"hidden": 1
		},
		{
			"fieldname":"group_by",
			"label": __("Group by"),
			"fieldtype": "Select",
			"options": ["", __("Group by Voucher"), __("Group by Voucher (Consolidated)"),
				__("Group by Account"), __("Group by Party")],
			"default": __("Group by Voucher (Consolidated)"),
			"hidden": 1
		},
		{
			"fieldname":"tax_id",
			"label": __("Tax Id"),
			"fieldtype": "Data",
			"hidden": 1
		},
		{
			"fieldname": "presentation_currency",
			"label": __("Currency"),
			"fieldtype": "Select",
			"options": erpnext.get_presentation_currency_list(),
			"default": "INR",
			"hidden": 1
		},
		{
			"fieldname":"cost_center",
			"label": __("Cost Center"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Cost Center', txt);
			},
			"hidden": 1
		},
		{
			"fieldname":"project",
			"label": __("Project"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Project', txt);
			},
			"hidden": 1
		},
		{
			"fieldname": "show_opening_entries",
			"label": __("Show Opening Entries"),
			"fieldtype": "Check",
			"hidden": 1
		},
		{
			"fieldname": "include_default_book_entries",
			"label": __("Include Default Book Entries"),
			"fieldtype": "Check",
			"default": 1,
			"hidden": 1
		},
		{
			"fieldname": "show_cancelled_entries",
			"label": __("Show Cancelled Entries"),
			"fieldtype": "Check",
			"hidden": 1
		}
	],
	onload: function(query_report) {
		query_report.page.clear_menu();
		query_report.page.add_menu_item(__("PDF"), () => {
			const filters = JSON.stringify(query_report.get_values());
			print_settings = {'orientation': 'Portrait'};
			frappe.call({
				method: "essdee.essdee.report.party_ledger.party_ledger.get_ledger_print_html",
				args: {ledger_filters: filters},
				callback: function(r) {
					if(r.message) {
						frappe.render_pdf(r.message, print_settings);
					}
				}
			}
		);
	}
		)
}
};
